src/model.cpp

```diff
@@ -339,4 +339,159 @@ bool Model::hasUnresolvedImports()
+{
+    size_t index = 0;
+    bool found = false;
+    while (index < componentParent->componentCount() && !found) {
+        if (componentParent->component(index) == component) {
+            found = true;
+        } else {
+            ++index;
+        }
+    }
+
+    return index;
+}
+
+IndexStack reverseEngineerIndexStack(const VariablePtr &variable)
+{
+    IndexStack indexStack;
+    ComponentPtr component = std::dynamic_pointer_cast<Component>(variable->parent());
+    indexStack.push_back(getVariableIndexInComponent(component, variable));
+
+    ComponentEntityPtr parent = component;
+    ComponentEntityPtr grandParent = std::dynamic_pointer_cast<ComponentEntity>(parent->parent());
+    while (grandParent != nullptr) {
+        indexStack.push_back(getComponentIndexInComponentEntity(grandParent, parent));
+        parent = grandParent;
+        grandParent = std::dynamic_pointer_cast<ComponentEntity>(grandParent->parent());
+    }
+
+    std::reverse(std::begin(indexStack), std::end(indexStack));
+
+    return indexStack;
+}
+
+void recordVariableEquivalences(const ComponentPtr &component, EquivalenceMap &equivalenceMap, IndexStack &indexStack)
+{
+    for (size_t index = 0; index < component->variableCount(); ++index) {
+        auto variable = component->variable(index);
+        for (size_t j = 0; j < variable->equivalentVariableCount(); ++j) {
+            if (j == 0) {
+                indexStack.push_back(index);
+            }
+            auto equivalentVariable = variable->equivalentVariable(j);
+            auto equivalentVariableIndexStack = reverseEngineerIndexStack(equivalentVariable);
+            if (equivalenceMap.count(indexStack) == 0) {
+                equivalenceMap[indexStack] = std::vector<IndexStack>();
+            }
+            equivalenceMap[indexStack].push_back(equivalentVariableIndexStack);
+        }
+        if (variable->equivalentVariableCount() > 0) {
+            indexStack.pop_back();
+        }
+    }
+}
+
+void generateEquivalenceMap(const ComponentPtr &component, EquivalenceMap &map, IndexStack &indexStack)
+{
+    for (size_t index = 0; index < component->componentCount(); ++index) {
+        indexStack.push_back(index);
+        auto c = component->component(index);
+        recordVariableEquivalences(c, map, indexStack);
+        generateEquivalenceMap(c, map, indexStack);
+        indexStack.pop_back();
+    }
+}
+
+EquivalenceMap generateEquivalenceMap(const std::shared_ptr<const libcellml::Model> &model)
+{
+    EquivalenceMap map;
+
+    IndexStack indexStack;
+    for (size_t index = 0; index < model->componentCount(); ++index) {
+        indexStack.push_back(index);
+        auto c = model->component(index);
+        recordVariableEquivalences(c, map, indexStack);
+        generateEquivalenceMap(c, map, indexStack);
+        indexStack.pop_back();
+    }
+
+    return map;
+}
```

Comment:
This method is not needed as such. Its body could be directly in `Model::clone()`.

Reply:
In fact, `Model::clone()` could declare `map` and `indexStack`, and then call a generic `generateEquivalenceMap()` (since both `Model` and `Component` have components)? This would avoid duplicated code.

Reply:
The problem is the `shared_from_this` it is a subtly different type of shared pointer which can't be converted to the shared pointers that we typically use.

Reply:
Ok, I thought it might have been a problem. Fair enough then re avoiding duplication of code. However, we might still want to move the contents of the method to `Model::clone()`. Not critical though.